### PR TITLE
Unimodal thresh upsampling

### PR DESCRIPTION
**Pull request by anithapk**
Additions:
- unimodal thresholding
- resizing volumes
:-)

### CODE REVIEW

File: micro_dl/utils/image_utils.py [hunk 1]
```diff
@@ -190,3 +219,69 @@ def read_image(file_path):
...
+
+
+def get_unimodal_threshold(input_image):
+    """Determines optimal unimodal threshold
+
+    https://users.cs.cf.ac.uk/Paul.Rosin/resources/papers/unimodal2.pdf
+    https://www.mathworks.com/matlabcentral/fileexchange/45443-rosin-thresholding
+
+    :param np.array input_image: generate mask for this image
+    :return float best_threshold: optimal lower threshold for the fg hist
```
Comment (jennyfolkesson) [inline micro_dl/utils/image_utils.py]: I'm not sure it's obvious what fg is, would you mind expanding on "fg hist"?

File: micro_dl/utils/image_utils.py [hunk 2]
```diff
@@ -190,3 +219,69 @@ def read_image(file_path):
...
+        y0 = hist_counts[idx]
+        a = [p1[0] - p2[0], p1[1] - p2[1]]
+        b = [x0 - p2[0], y0 - p2[1]]
+        cross_ab = a[0] * b[1] - b[0] * a[1]
+        per_dist = np.linalg.norm(cross_ab) / np.linalg.norm(a)
+        if per_dist > max_dist:
+            best_threshold = x0
+            max_dist = per_dist
+
+    return best_threshold
```
Comment (jennyfolkesson) [inline micro_dl/utils/image_utils.py]: Would there ever be a histogram where best_threshold stays -1 (e.g. max_idx == last_binedge)? If so, maybe do a final check that best_threshold is > -1?

File: micro_dl/utils/mp_utils.py [hunk 3]
```diff
@@ -258,3 +258,52 @@ def resize_and_save(**kwargs):
...
+        meta_idx = aux_utils.get_meta_idx(frames_metadata,
+                                          time_idx,
+                                          channel_idx,
+                                          sl_idx,
+                                          pos_idx)
+        cur_fname = frames_metadata.loc[meta_idx, 'file_name']
+        cur_img = image_utils.read_image(os.path.join(input_dir, cur_fname))
+        input_stack.append(cur_img)
+    input_stack = np.stack(input_stack, axis=0)
+    resc_vol = image_utils.rescale_nd_image(input_stack, scale_factor)
```
Comment (jennyfolkesson) [inline micro_dl/utils/mp_utils.py]: I was just wondering - how will these resizes volumes work with the remainder of the preprocessing - flatfield, masks and tiles (which I think assumes 2D frames)?

File: micro_dl/preprocessing/resize_images.py [hunk 4]
```diff
@@ -107,3 +113,66 @@ def resize_frames(self):
...
+                                        'channel_idx': channel_idx,
+                                        'slice_start_idx': start_idx,
+                                        'slice_end_idx': end_idx,
+                                        'file_name': op_fname}
+                        resized_metadata_list.append(cur_metadata)
+
+        # Multiprocessing of kwargs
+        mp_utils.mp_rescale_vol(mp_args, self.num_workers)
+        resized_metadata_df = pd.DataFrame.from_dict(resized_metadata_list)
+        resized_metadata_df.to_csv(
```
Comment (jennyfolkesson) [inline micro_dl/preprocessing/resize_images.py]: Won't there be a problem when frames_meta has different fields than expected in the rest of the processing steps like tiling and masks?

File: micro_dl/preprocessing/resize_images.py [hunk 5]
```diff
@@ -107,3 +113,66 @@ def resize_frames(self):
...
+        if num_slices_subvolume == -1:
+            num_slices_subvolume = len(self.slice_ids)
+        num_blocks = np.ceil(
+            num_total_slices / num_slices_subvolume
+        ).astype('int')
+        for time_idx in self.time_ids:
+            for pos_idx in self.pos_ids:
+                for channel_idx in self.channel_ids:
+                    for slice_idx in range(num_blocks):
+                        start_idx = slice_idx * num_slices_subvolume
```
Comment (jennyfolkesson) [inline micro_dl/preprocessing/resize_images.py]: When there's no overlap in z between blocks, will there be discontinuities between the upsampled blocks?

File: micro_dl/cli/preprocess_script.py [hunk 6]
```diff
@@ -82,8 +82,8 @@ def pre_process(pp_config):
         num_workers = pp_config['num_workers']
 
     # Resample images first
-    if 'resample_scale' in pp_config:
-        scale_factor = pp_config['resample_scale']
+    if 'resize' in pp_config:
```
Comment (jennyfolkesson) [inline micro_dl/cli/preprocess_script.py]: Would you mind adding description expected resize parameters shape and order in the readme?

File: micro_dl/utils/image_utils.py (hunk 2 above)
Comment (anithapk) [reply micro_dl/utils/image_utils.py]: good point! The images are uint16 and not normalized. After flat field correction, they are float but with non negative values.  Changing best_threshold = np.inf, max_dist = -np.inf and adding an assert before return!

File: micro_dl/utils/mp_utils.py (hunk 3 above)
Comment (anithapk) [reply micro_dl/utils/mp_utils.py]: I've been debating the same! The options I'm thinking are:
1) after resizing, we could save it as 2D slices and the slice_idx will correspond to the center slice but tiling has to be modified to allow even number of slices
2) Save it as 3D volume where slice_idx will correspond to the starting slice in sub volume and use tile 3D option
I've a slight preference for 2 but not too attached to it!

File: micro_dl/preprocessing/resize_images.py (hunk 4 above)
Comment (anithapk) [reply micro_dl/preprocessing/resize_images.py]: very true! I'll replace slice_start_idx with start_idx and remove slice_end_idx. We should be able to get the number of tiles per subvolume from the config/json.

File: micro_dl/preprocessing/resize_images.py (hunk 5 above)
Comment (anithapk) [reply micro_dl/preprocessing/resize_images.py]: Yes, there will be a discontinuity of one slice! Thanks for catching this, the index here is also wrong (assumes slice_idx starting from 0)....

**Pull request closed without merging**